src/gui/URLEdit.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ *  Copyright (C) 2014 Felix Geyer <debfx@fobos.de>
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "URLEdit.h"
+
+#include <QRegularExpression>
+
+#include "core/Config.h"
+#include "core/FilePath.h"
+#include "core/Tools.h"
+#include "gui/Font.h"
+
+const QColor URLEdit::CorrectSoFarColor = QColor(255, 205, 15);
+const QColor URLEdit::ErrorColor = QColor(255, 125, 125);
+
+URLEdit::URLEdit(QWidget* parent)
+    : QLineEdit(parent)
+{
+    const QIcon errorIcon = filePath()->icon("status", "dialog-error");
+    m_errorAction = addAction(errorIcon, QLineEdit::TrailingPosition);
+    m_errorAction->setVisible(false);
+    m_errorAction->setToolTip(tr("Invalid URL"));
+
+    const QIcon correctIcon = filePath()->icon("actions", "dialog-ok");
+    m_correctAction = addAction(correctIcon, QLineEdit::TrailingPosition);
+    m_correctAction->setVisible(false);
+    m_correctAction->setToolTip(tr("URL is valid"));
+
+    updateStylesheet();
+}
+
+void URLEdit::enableVerifyMode()
+{
+    updateStylesheet();
+
+    connect(this, SIGNAL(textChanged(QString)), SLOT(updateStylesheet()));
+}
+
+void URLEdit::updateStylesheet()
+{
+    const QString stylesheetTemplate("QLineEdit { background: %1; }");
+
+    if (!Tools::validURL(text())) {
+        setStyleSheet(stylesheetTemplate.arg(ErrorColor.name()));        
+        m_correctAction->setVisible(false);
+        m_errorAction->setVisible(true);
+    } else {
+        m_correctAction->setVisible(false);
```

Comment:
Shouldn't this be `setVisible(true)`? I don't see `m_correctAction` being made visible anywhere.

Reply:
Ah, actually I thought that we don't need toe correct action at all. It's originally only needed when the passwords match. I guess we don't want to always show a confirmation in the entry URL field if it's correct. Only show the invalid value. So the correct action member could be actually removed.

Reply:
Yep, I was thinking the same! 

Another thing I wanted to say is that I think this validation shouldn't actually prevent users from putting in a malformed URL, it should just be a warning, but it should be ignorable. (I didn't actually see any logic that forbids an incorrect URL, I just wanted to make my opinion explicit)

Reply:
You understood correctly, this is just a warning. It doesn't prevent using it. Still, of course it won't work properly.